core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala

```diff
@@ -0,0 +1,232 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.util.collection
+
+import java.io.File
+import java.util.Comparator
+import java.util.concurrent.{PriorityBlockingQueue, CountDownLatch}
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark._
+import org.apache.spark.executor.ShuffleWriteMetrics
+import org.apache.spark.storage.BlockId
+import org.apache.spark.serializer.Serializer
+import org.apache.spark.util.CompletionIterator
+
+/**
+ * Manages blocks of sorted data on disk that need to be merged together. Carries out a tiered
+ * merge that will never merge more than spark.shuffle.maxMergeFactor segments at a time.
+ * Except for the final merge, which merges disk blocks to a returned iterator, TieredDiskMerger
+ * merges blocks from disk to disk.
+ *
+ * TieredDiskMerger carries out disk-to-disk merges in a background thread that can run concurrently
+ * with blocks being deposited on disk.
+ *
+ * When deciding which blocks to merge, it first tries to minimize the number of blocks, and then
+ * the size of the blocks chosen.
+ */
+private[spark] class TieredDiskMerger[K, C](
+    conf: SparkConf,
+    dep: ShuffleDependency[K, _, C],
+    keyComparator: Comparator[K],
+    context: TaskContext) extends Logging {
+
+  /** Manage the on-disk shuffle block and related file, file size */
+  case class DiskShuffleBlock(blockId: BlockId, file: File, len: Long)
+      extends Comparable[DiskShuffleBlock] {
+    def compareTo(o: DiskShuffleBlock): Int = len.compare(o.len)
+  }
+
+  private val maxMergeFactor = conf.getInt("spark.shuffle.maxMergeFactor", 100)
+  private val fileBufferSize = conf.getInt("spark.shuffle.file.buffer.kb", 32) * 1024
+
+  private val blockManager = SparkEnv.get.blockManager
+  private val ser = Serializer.getSerializer(dep.serializer)
+
+  /** PriorityQueue to store the on-disk merging blocks, blocks are merged by size ordering */
+  private val onDiskBlocks = new PriorityBlockingQueue[DiskShuffleBlock]()
+
+  /** A merging thread to merge on-disk blocks */
+  private val diskToDiskMerger = new DiskToDiskMerger
+
+  /** Signal to block/signal the merge action */
+  private val mergeReadyMonitor = new AnyRef()
+
+  private val mergeFinished = new CountDownLatch(1)
+
+  /** Whether more on-disk blocks may come in */
+  @volatile private var doneRegistering = false
+
+ /** Number of bytes spilled on disk */
+  private var _diskBytesSpilled: Long = 0L
+
+  def diskBytesSpilled: Long = _diskBytesSpilled
+
+  def registerOnDiskBlock(blockId: BlockId, file: File): Unit = {
+    assert(!doneRegistering)
+    onDiskBlocks.put(new DiskShuffleBlock(blockId, file, file.length()))
+
+    mergeReadyMonitor.synchronized {
+      if (shouldMergeNow()) {
+        mergeReadyMonitor.notify()
+      }
+    }
+  }
+
+  /**
+   * Notify the merger that no more on disk blocks will be registered.
+   */
+  def doneRegisteringOnDiskBlocks(): Unit = {
+    doneRegistering = true
```

Comment:
there is a dead lock with Line 176. so i think we need to put doneRegistering = true to mergeReadyMonitor.synchronized {}.